internal/terraform/eval_context_builtin.go

```diff
@@ -498,7 +498,7 @@ func (ctx *BuiltinEvalContext) evaluationExternalFunctions() lang.ExternalFuncs
 		ret.Provider[localName] = make(map[string]function.Function, len(funcDecls))
 		funcs := ret.Provider[localName]
 		for name, decl := range funcDecls {
-			funcs[name] = decl.BuildFunction(name, func() (providers.Interface, error) {
+			funcs[name] = decl.BuildFunction(providerAddr, name, func() (providers.Interface, error) {
```

Comment:
Since it's `EvalContext` that's calling `BuildFunction`, could we have the "function results cache" live as a member of `EvalContext` instead of a global, and then be passed in to `BuildFunction` here?

I'm nervous about adding a global table that just keeps growing over time. Although in a `terraform plan` command we can be confident that it'll only grow to the extent required for that particular plan, creating plans through a long-running `terraform rpcapi` process would presumably cause that global table to potentially grow for each call (unless the function calls are identical between plans).

Reply:
Thanks, `rpcapi` slipped my mind. That does seem like a good place to slip in a more local state for this, let me move things around a bit.

Reply:
Other than shear simplicity, one of the driving factors for the package-level cache was that it can be effective to carry the state from `Validate` forward into the plan. Most functions calls within a plan will only be called once, and won't fail until are partway through the apply step, and having a record from a prior call in validate give us another chance to fail early. In order to do that however we need to step not just up to the core `Context`, but all the way out to the local backend and carry the data through the `ContextOpts` (where we have coincidentally removed transferred program state, keeping it all the config and plan).

I'm not sure though _how_ effective that is in practice, since a lot of data during Validate is unknown, and maybe in real worlds scenarios most calls during validation wouldn't be made with real data anyway. 

Having the cache live only in the run context pollutes enough of the core API already, I don't like the idea of it requiring special handling outside of core in the backend too. What do you think about dropping the data from `Validate` and just relying on plan+apply alone?

Reply:
By this, do you mean designing it so that walks starting from `Context.Validate` would just not populate the results table _at all_? So we'd build the table only during the plan phase, and refer to the table only during the apply phase.

If that _is_ what you intended then that seems intuitive to me: we're building this table only to verify that providers are keeping promises made during planning. Validation doesn't yield any result other than diagnostics, so if a provider somehow does something inconsistent during validate then it might cause the validation results to be inconsistent, but the plan phase performs various validation-like checks itself anyway and so disagreement between validate and plan is not such a big deal.

Reply:
Thanks, I'm going to push up a version that's contained only within the `terraform` package for now then.

Just to clarify,, `Validate` would still have an active table for validation, and if there were multiple calls with matching arguments during the course of a single `Validate` call it would catch inconsistent results. It's just kind of awkward to transfer those records from `Validate` to `Plan`, because they are only procedurally joined in the backend.

Reply:
Ahh, I see... I wasn't considering the case where there are two identical calls to the same function in different parts of the configuration, but given that it does seem nice to be able to catch inconsistencies between those results during a single validate walk, even though the constraint would not then carry forward directly into the plan walk.